### ISSUE TITLE
[GlobalStep] "Block validation" error is displayed in the console after navigating to post/page with Filter by Attributes, Filter by Stock, Filter by Rating and Products (beta) block.

### ISSUE THREAD

**Issue by gglobalstep** [woothemes/woocommerce#39358]
### Prerequisites

- [ ] I have carried out troubleshooting steps and I believe I have found a bug.
- [ ] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

"Block validation" error is displayed in the console after navigating to post/page with Filter by Attributes, Filter by Stock, Filter by Rating and Products (beta) block.

### Expected behavior

Post/page with Filter by Attributes, Filter by Stock, Filter by Rating and Products (beta) block. should be shown without any error.

### Actual behavior

"Block validation" error is displayed in the console after navigating to post/page with Filter by Attributes, Filter by Stock, Filter by Rating and Products (beta) block.

### Steps to reproduce

1. Create any test site.
2. Install and activate all the required plugins.
3. Install and activate the release Woocommerce plugin (i.e. v7.9.0) from the store.
4. Add post/page and test the following blocks Filter by Attributes, Filter by Stock Filter by Rating and Products (beta) block. Just leave all default settings and publish post/page.
5. Install and activate Woocommerce 8.0.0-beta.1 plugin.
6. Open the post/page you created in Step 1 
7. Observe that, "Block validation" error is displayed in the console after navigating to post/page with Filter by Attributes, Filter by Stock, Filter by Rating and Products (beta) block.

### Video:

https://github.com/woocommerce/woocommerce/assets/41110392/bc160230-a0e7-4b60-8c0d-a518bdb7a207

### WordPress Environment

Woocommerce Version: 8.0.0-beta.1
WordPress version: 6.2.2
PHP version: 8.0.29

PC:
Windows 10
Chrome(Version 115.0.5790.99)
Firefox (Version 115.0.2)

### Isolating the problem

- [ ] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [ ] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

<details>

`
### WordPress Environment ###

WC Version: 8.0.0
REST API Version: ✔ 8.0.0
WC Blocks Version: ✔ 10.6.0
Action Scheduler Version: ✔ 3.6.1
Log Directory Writable: ✔
WP Version: 6.2.2
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: ✔
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: Apache/2.4.57 (Unix) OpenSSL/1.0.2g
PHP Version: 8.0.29
PHP Post Max Size: 1 GB
PHP Time Limit: 30
PHP Max Input Vars: 5000
cURL Version: 7.47.0
OpenSSL/1.0.2g

SUHOSIN Installed: –
MySQL Version: 5.7.33-0ubuntu0.16.04.1-log
Max Upload Size: 512 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 7.9.0
WC Database Prefix: wp_
Total Database Size: 5.40MB
Database Data Size: 3.67MB
Database Index Size: 1.73MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_order_itemmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_actions: Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wp_actionscheduler_claims: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_groups: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_logs: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_comments: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_options: Data: 2.47MB + Index: 0.06MB + Engine InnoDB
wp_postmeta: Data: 0.11MB + Index: 0.06MB + Engine InnoDB
wp_posts: Data: 0.08MB + Index: 0.06MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_admin_notes: Data: 0.06MB + Index: 0.00MB + Engine InnoDB
wp_wc_admin_note_actions: Data: 0.05MB + Index: 0.02MB + Engine InnoDB
wp_wc_category_lookup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_customer_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_orders: Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wp_wc_orders_meta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_addresses: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_coupon_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_operational_data: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_product_lookup: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_stats: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_order_tax_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_product_attributes_lookup: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_download_directories: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_rate_limits: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_reserved_stock: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB

### Post Type Counts ###

attachment: 24
page: 9
post: 5
product: 18
product_variation: 7
revision: 1
wp_global_styles: 1
wp_template: 2

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (3) ###

Query Monitor: by John Blackbourn – 3.13.1
Companion Plugin: by Osk – 1.29
WooCommerce: by Automattic – 8.0.0-beta.1

### Inactive Plugins (3) ###

Akismet Anti-Spam: Spam Protection: by Automattic - Anti Spam Team – 5.2
Hello Dolly: by Matt Mullenweg – 1.7.2
Jetpack: by Automattic – 12.3

### Dropin Plugins (1) ###

db.php: Query Monitor Database Class (Drop-in)

### Settings ###

API Enabled: –
Force SSL: –
Currency: USD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: –
Enforce Approved Product Download Directories: ✔
HPOS feature screen enabled: –
HPOS feature enabled: –
Order datastore: WC_Order_Data_Store_CPT
HPOS data sync enabled: –

### WC Pages ###

Shop base: #8 - /shop/
Cart: #9 - /cart/
Checkout: #10 - /checkout/
My account: #11 - /my-account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: Twenty Twenty-Three
Version: 1.1
Author URL: https://wordpress.org
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ✔

### Templates ###

Overrides: –

### Admin ###

Enabled Features: activity-panels
analytics
product-block-editor
coupons
core-profiler
customer-effort-score-tracks
import-products-task
experimental-fashion-sample-products
shipping-smart-defaults
shipping-setting-tour
homescreen
marketing
mobile-app-banner
navigation
onboarding
onboarding-tasks
remote-inbox-notifications
remote-free-extensions
payment-gateway-suggestions
shipping-label-banner
subscriptions
store-alerts
transient-notices
woo-mobile-welcome
wc-pay-promotion
wc-pay-welcome-page

Disabled Features: minified-js
new-product-management-experience
product-variation-management
settings
async-product-editor-category-field

Daily Cron: ✔ Next scheduled: 2023-07-22 07:31:58 +00:00
Options: ✔
Notes: 65
Onboarding: skipped

### Action Scheduler ###

Complete: 31
Oldest: 2023-07-21 07:33:03 +0000
Newest: 2023-07-21 07:44:00 +0000

Failed: 1
Oldest: 2023-07-21 07:33:03 +0000
Newest: 2023-07-21 07:33:03 +0000

Pending: 1
Oldest: 2023-07-22 07:33:03 +0000
Newest: 2023-07-22 07:33:03 +0000

### Status report information ###

Generated at: 2023-07-21 09:38:11 +00:00
`

</details>

**Comment by kmanijak** [comment]
Related to https://github.com/woocommerce/woocommerce-blocks/issues/10298

**Comment by nefeline** [comment]
> Related to https://github.com/woocommerce/woocommerce-blocks/issues/10298

Why do you believe these issues are related @kmanijak ? I don't see any relationship between what is reported here and the problem on `#10298`, which is simply a mismatch between pattern styles in the editor and on the FE.

**Comment by kmanijak** [comment]
> Why do you believe these issues are related @kmanijak ?

The error we get is exactly the same, block validation for `woocommerce/product-button` fails with the same difference between save function and actual output. The way of triggering the error is different, but I assume there’s high chance they have the same reason

**Comment by kmanijak** [comment]
> Related to https://github.com/woocommerce/woocommerce-blocks/issues/10298
> The error we get is exactly the same, block validation for woocommerce/product-button fails with the same difference between save function and actual output. The way of triggering the error is different, but I assume there’s high chance they have the same reason.

It's not valid, I made an incorrect assumption. Apologies for the confusion.

**Comment by nefeline** [comment]
Hey @kmanijak! Following up on this, I did some additional tests and was able to find an instance where there is indeed a relationship between this error and the one on https://github.com/woocommerce/woocommerce-blocks/issues/10298, even though it's still unclear why this problem is happening. As it turns out:

* This error `Block validation: Block validation failed for woocommerce/product-button`, happens when running the woocommerce-blocks `10.6.0` bundled in the core of Woo (which is what we have on version 8.0.0-beta.1), but it doesn’t happen when having Woo Blocks activated as a separate plugin on this exact same version `10.6.0`. This explains why we were both getting mixed results depending on how the test environment was setup.
* This error is not at all associated with the block filters as originally reported on this issue: it is triggered whenever the editor is accessed, independent of which blocks are being used (only with the bundled version of Woo Blocks in the core of Woo, not the standalone plugin).
* The removal of [opinionated styles from the Product Hero pattern](https://github.com/woocommerce/woocommerce-blocks/pull/10255) solves the error in the core of Woo, but doesn't explain why the error is **never triggered** in our standalone plugin running on the exact same version `10.6.0`.

So to summarize, as part of this issue, we still need to figure out what could cause a difference between the version of `woocommerce-blocks` bundled within WooCommerce and the same version running as a standalone plugin as this shakes our confidence in developing new patterns: how can we ensure our patterns are fully functional if errors such as this one are not triggered during development and in our standalone plugin, but triggered in the core of Woo?

**Comment by Aljullu** [comment]
I have been investigating this as well, and this is what I've found so far:

* As @nefeline mentioned, the issue is only reproducible with the built versions of Blocks in WC core. It can't be reproduced in the feature plugin. These are some of the environments that I tested:
  * WooCommerce 8.0.0-beta.1 → reproducible
  * WooCommerce 8.0.0-beta.1 + WC Blocks trunk → not reproducible
  * WooCommerce trunk → reproducible
  * WooCommerce trunk + WC Blocks trunk → not reproducible
* Based on that, I thought the issue might be related to some `isFeaturePlugin()` check somewhere, but wasn't able to find any that would impact it.
* The issue is reproducible independently of the theme. Using a block or classic theme doesn't make any difference.
* Having Gutenbeg enabled or disabled doesn't make any difference either, neither testing on WP 6.2.2 or 6.3-RC1.
  * On WP 6.3-RC1 there is also a validation error for the `core/image` block, but that seems to be a [different issue](https://github.com/WordPress/gutenberg/issues/52388#issuecomment-1644926375).
* Taking a look at the error, it seems like some styles are being added to the save function: `color`, `background-color`, `margin-top`, `padding-right`, `padding-left`.

```diff
- <div class="wp-block-woocommerce-product-button is-loading" style="font-size:15px"></div>
+ <div class="wp-block-woocommerce-product-button is-loading has-text-color has-background" style="color:#fffff1;background-color:#000001;margin-top:15px;padding-right:30px;padding-left:30px;font-size:15px"></div>
```

After digging a bit more into the Gutenberg and WooCommerce code codebase... I think the issue comes from the version of the Product Hero pattern included in WC core, which had those styles:

https://github.com/woocommerce/woocommerce-blocks/pull/10255/files#diff-11677bab323ab51ec7de5d0784fd13f2328316839352bc2c2a16d26f0f0b8ec9L36

So if my theory is right, this has been fixed in https://github.com/woocommerce/woocommerce-blocks/pull/10255. I will try again tomorrow and see if that's the correct hypothesis.

**Comment by Aljullu** [comment]
I'm going to close this issue based on my comment above. To verify, you can remove the `product-hero.php` pattern from `/packages/woocommerce-blocks/patterns/` and you will see that the issue is gone. The issue happens because the pattern uses some [styles which are not enabled in WC core](https://github.com/woocommerce/woocommerce-blocks/blob/b4ba06a6242cbb6a64d2ec554a263ebe60d8d3af/assets/js/atomic/blocks/product-elements/button/index.tsx#L17), so the block can't be validated properly. As mentioned, this was fixed in https://github.com/woocommerce/woocommerce-blocks/pull/10255.

**Comment by nefeline** [comment]
> So if my theory is right, this has been fixed in https://github.com/woocommerce/woocommerce-blocks/pull/10255. I will try again tomorrow and see if that's the correct hypothesis.

Yep, it fixes the problem indeed, I did the tests to confirm & shared the results in my previous comment here: https://github.com/woocommerce/woocommerce/issues/39358#issuecomment-1646806072

> The removal of [opinionated styles from the Product Hero pattern](https://github.com/woocommerce/woocommerce-blocks/pull/10255) solves the error in the core of Woo, but doesn't explain why the error is never triggered in our standalone plugin running on the exact same version 10.6.0.

Regarding:

> The issue happens because the pattern uses some [styles which are not enabled in WC core](https://github.com/woocommerce/woocommerce-blocks/blob/b4ba06a6242cbb6a64d2ec554a263ebe60d8d3af/assets/js/atomic/blocks/product-elements/button/index.tsx#L17), so the block can't be validated properly.

Nice finding! This a great point to keep in mind when developing new opinionated patterns (and in a way, it's a good thing we are opting for wireframed now).
